### PR TITLE
Require empty lines after subject and latest let

### PR DESCRIPTION
**Pull request by Darhazer**

### CODE REVIEW

**Comment by backus** [comment]
Nice! A few comments:

* I love this kind of cop and I've really been wanting something like this. I think this cop should add autocorrect though before we merge it. My private codebase I ran it on is pretty polished and it found 66 offenses. Manually correcting a bunch of newlines is a pain

* I think I found some false positives. Try testing this case:
    ```ruby
    RSpec.shared_context 'foo' do
      let(:bar) { baz }
      let(:baz) { qux }
    end

    RSpec.describe Foo do
      it_beahves_like 'x' do
        let(:bar) { baz }
        let(:baz) { qux }
      end
    end
    ```

**Comment by bquorning** [comment]
I have identical observations. Ran the cops on some 1200 files, got 105 offenses. A few of them were false positives where `let` is the last line in a block.

**Comment by Darhazer** [comment]
Added autocorrect
Fixed the false positives
Added changelog entries

File: spec/rubocop/cop/rspec/empty_line_after_let_spec.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+  it 'ignores empty lines between the lets' do
+    expect_violation(<<-RUBY)
+      RSpec.describe User do
+        let(:a) { a }
+
+        subject { described_class }
+
+        let!(:b) { b }
+        ^^^^^^^^^^^^^^ Add empty line after the last `let` block.
```
Comment (bquorning) [inline spec/rubocop/cop/rspec/empty_line_after_let_spec.rb]: “… and group all `let` lines together” – but that’s for another cop :-)

File: lib/rubocop/cop/rspec/empty_line_after_let.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      #
+      #   # good
+      #     let(:foo) { bar }
+      #     let(:something) { other }
+      #
+      #     it do
+      #       ...
+      #     end
+      class EmptyLineAfterLet < Cop
+        MSG = 'Add empty line after the last `let` block.'.freeze
```
Comment (backus) [inline lib/rubocop/cop/rspec/empty_line_after_let.rb]: Can you change this to `"Add an empty line"` so it uses proper grammar?

**Comment by backus** [comment]
I think we should probably rename `EmptyLineAfterLet` to `EmptyLineAfterFinalLet`

**Comment by backus** [comment]
Wow. I ran this cop on a few codebases I work with an it did a great job! This is the kind of cop I've wanted for a long time I'm glad you put in the time and energy :smile:

**Comment by backus** [comment]
@Darhazer I have another small bug for you:

```
$ cat ex_spec.rb
RSpec.describe 'silly heredoc syntax' do
  let(:foo) { <<-BAR }
  hello
  world
  BAR

  it 'has tricky syntax' do
    expect(foo).to eql("  hello\n  world\n")
  end
end

$ bundle exec rubocop --require rubocop-rspec --only RSpec/EmptyLineAfterLet ex_spec.rb
Inspecting 1 file
C

Offenses:

ex_spec.rb:2:3: C: RSpec/EmptyLineAfterLet: Add empty line after the last let block.
  let(:foo) { <<-BAR }
  ^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected
-> [1]

$ bundle exec rubocop --require rubocop-rspec --only RSpec/EmptyLineAfterLet -a ex_spec.rb
Inspecting 1 file
C

Offenses:

ex_spec.rb:2:3: C: [Corrected] RSpec/EmptyLineAfterLet: Add empty line after the last let block.
  let(:foo) { <<-BAR }
  ^^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected, 1 offense corrected

$ cat ex_spec.rb
RSpec.describe 'silly heredoc syntax' do
  let(:foo) { <<-BAR }

  hello
  world
  BAR

  it 'has tricky syntax' do
    expect(foo).to eql("  hello\n  world\n")
  end
end
```

I found this while running your cops on the main rubocop codebase

**Comment by backus** [comment]
@Darhazer I'm not sure if there are others but I recommend checking by running this on the rubocop codebase when you fix this bug. It is a good way to fuzz your implementation and it also can become a very easy and fun PR you can offer up to the rubocop team when you run autocorrect :smile:

File: spec/rubocop/cop/rspec/empty_line_after_let_spec.rb (hunk 1 above)
Comment (Darhazer) [inline spec/rubocop/cop/rspec/empty_line_after_let_spec.rb]: Nice idea. I will work on this next ;)

**Comment by Darhazer** [comment]
PR updated

**Comment by Darhazer** [comment]
Tested on rubocop and one more code base for false positives.
Is there anything else to be added/changed here?

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,9 @@
 
 ## Master (Unreleased)
 
+* Add `RSpec/EmptyLineAfterSubject` cop. ([@Darhazer][])
+* Add `RSpec/EmptyLineAfterLet` cop. ([@Darhazer][])
```
Comment (backus) [inline CHANGELOG.md]: This should probably also say `EmptyLineAfterFinalLet`

**Comment by backus** [comment]
@Darhazer these cops are great! Sorry it took me a bit longer to get to the final reviews. If you could just update the cop name in the changelog you can push and then merge assuming CI passes :smile:

**Comment by backus** [comment]
Great work @Darhazer!

**Pull request merged**